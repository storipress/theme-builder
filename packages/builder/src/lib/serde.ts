import { parse, stringify } from 'devalue'
import { isReactive, isRef, isShallow, reactive, ref, shallowRef, toRaw } from 'vue'

const serializeReducers = {
  Reactive: (obj: unknown) => {
    if (isReactive(obj)) {
      // vue 2 is same reference
      return structuredClone(obj)
    }
  },
  Ref: (obj: unknown) => {
    if (isRef(obj) && !isShallow(obj)) {
      return toRaw(obj.value)
    }
  },
  ShallowRef: (obj: unknown) => {
    if (isRef(obj) && isShallow(obj)) {
      return toRaw(obj.value)
    }
  },

  FreezedArray: (obj: unknown) => {
    if (typeof obj === 'object' && Object.isFrozen(obj) && Array.isArray(obj)) {
      return [...obj]
    }
  },
  FreezedObject: (obj: unknown) => {
    if (typeof obj === 'object' && Object.isFrozen(obj) && !Array.isArray(obj) && obj !== null) {
      return {
        ...obj,
      }
    }
  },
}

export function serialize(obj: unknown) {
  return stringify(obj, serializeReducers)
}

const deserializeReducers = {
  Reactive: (obj: unknown) => {
    return reactive(obj as Record<string, unknown>)
  },
  Ref: (obj: unknown) => {
    return ref(obj)
  },
  ShallowRef: (obj: unknown) => {
    return shallowRef(obj)
  },
  FreezedArray: (obj: unknown) => {
    return Object.freeze(obj)
  },
  FreezedObject: (obj: unknown) => {
    return Object.freeze(obj)
  },
}

export function deserialize<T>(json: string): T {
  return parse(json, deserializeReducers)
}
