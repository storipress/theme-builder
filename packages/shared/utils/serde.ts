import { identity, isObject } from 'lodash'

interface SpecialObject {
  $$type: string
  value: unknown
}

function isSpecialObject(x: unknown): x is SpecialObject {
  if (isObject(x) && '$$type' in x) {
    return true
  }
  return false
}

export function serialize(obj: unknown) {
  return JSON.stringify(obj, (_, value) => {
    if (value instanceof Map) {
      return {
        $$type: 'map',
        value: [...value.entries()],
      }
    }
    if (value instanceof Set) {
      return {
        $$type: 'set',
        value: [...value.values()],
      }
    }
    if (Array.isArray(value) && Object.isFrozen(value)) {
      return {
        $$type: 'freezeArray',
        value: [...value],
      }
    }
    if (isObject(value) && Object.isFrozen(value)) {
      return {
        $$type: 'freeze',
        value: { ...value },
      }
    }
    return value
  })
}

function deserializeAs(Ctor: new (x: any) => unknown): (x: unknown) => unknown {
  return (x) => new Ctor(x)
}

const deserializeSpecialObject: Record<string, (x: unknown) => unknown> = {
  map: deserializeAs(Map),
  set: deserializeAs(Set),
  freeze: Object.freeze,
  freezeArray: Object.freeze,
}

export function deserialize<T>(json: string): T {
  return JSON.parse(json, (_, value) => {
    if (isSpecialObject(value)) {
      const des = deserializeSpecialObject[value.$$type] ?? identity
      return des(value.value)
    }
    return value
  })
}
