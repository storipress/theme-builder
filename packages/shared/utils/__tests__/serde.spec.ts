import { deserialize, serialize } from '../serde'

describe('serialize', () => {
  it('handle Map and Set', () => {
    expect(serialize({ map: new Map([['a', 'b']]), set: new Set(['a', 'b']) })).toMatchSnapshot()
  })

  it('handle freezed object', () => {
    expect(serialize({ freeze: Object.freeze({ foo: 'bar' }) })).toMatchSnapshot()
  })

  it('handle freezed array', () => {
    const json = serialize({ freeze: Object.freeze([1, 2, 3]) })
    expect(json).toMatchSnapshot()
  })
})

describe('deserialize', () => {
  it('handle Map and Set', () => {
    const json = serialize({ map: new Map([['a', 'b']]), set: new Set(['a', 'b']) })
    expect(deserialize(json)).toMatchSnapshot()
  })

  it('handle freezed object', () => {
    const json = serialize({ freeze: Object.freeze({ foo: 'bar' }) })
    const value = deserialize(json)
    expect(value).toMatchSnapshot()
    expect(Object.isFrozen((value as any).freeze)).toBe(true)
  })

  it('handle freezed array', () => {
    const json = serialize({ freeze: Object.freeze([1, 2, 3]) })
    const value = deserialize(json)
    expect(Object.isFrozen((value as any).freeze)).toBe(true)
    expect((value as any).freeze).toEqual([1, 2, 3])
  })
})
