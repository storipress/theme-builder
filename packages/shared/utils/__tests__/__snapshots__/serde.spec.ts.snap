// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`deserialize > handle Map and Set 1`] = `
{
  "map": Map {
    "a" => "b",
  },
  "set": Set {
    "a",
    "b",
  },
}
`;

exports[`deserialize > handle freezed object 1`] = `
{
  "freeze": {
    "foo": "bar",
  },
}
`;

exports[`serialize > handle Map and Set 1`] = `"{"map":{"$$type":"map","value":[["a","b"]]},"set":{"$$type":"set","value":["a","b"]}}"`;

exports[`serialize > handle freezed array 1`] = `"{"freeze":{"$$type":"freezeArray","value":[1,2,3]}}"`;

exports[`serialize > handle freezed object 1`] = `"{"freeze":{"$$type":"freeze","value":{"foo":"bar"}}}"`;
